Stratis.Bitcoin/BlockPulling/BlockPuller.cs

```diff
@@ -448,5 +356,279 @@ private void DistributeDownload(Dictionary<int, InventoryVector> vectors, BlockP
+
+        /// <summary>
+        /// Assigns a pending download task to a specific peer.
+        /// </summary>
+        /// <param name="peer">Peer to be assigned the new task.</param>
+        /// <param name="blockHash">If the function succeeds, this is filled with the hash of the block that will be requested from <paramref name="peer"/>.</param>
+        /// <returns>
+        /// <c>true</c> if a download task was assigned to the peer, <c>false</c> otherwise, 
+        /// which indicates that there was no pending task.
+        /// </returns>
+        internal bool AssignPendingDownloadTaskToPeer(BlockPullerBehavior peer, out uint256 blockHash)
+        {
+            blockHash = null;
+
+            lock (this.lockObject)
+            {
+                if (this.pendingInventoryVectors.Count > 0)
+                {
+                    blockHash = this.pendingInventoryVectors.Dequeue();
+                    this.map.Add(blockHash, peer);
+
+                    HashSet<uint256> peerPendingDownloads;
+                    if (!this.pendingDownloads.TryGetValue(peer, out peerPendingDownloads))
+                    {
+                        peerPendingDownloads = new HashSet<uint256>();
+                        this.pendingDownloads.Add(peer, peerPendingDownloads);
+                    }
+
+                    peerPendingDownloads.Add(blockHash);
+                }
+            }
+
+            bool res = blockHash != null;
+            return res;
+        }
+
+        /// <summary>
+        /// Assigns a download task to a specific peer.
+        /// </summary>
+        /// <param name="peer">Peer to be assigned the new task.</param>
+        /// <param name="blockHash">Hash of the block to download from <paramref name="peer"/>.</param>
+        /// <returns><c>true</c> if the block was assigned to the peer, <c>false</c> in case the block has already been assigned to someone.</returns>
+        /// <remarks>The caller of this method is responsible for holding <see cref="lockObject"/>.</remarks>
+        internal bool AssignDownloadTaskToPeer(BlockPullerBehavior peer, uint256 blockHash)
+        {
+            bool res = false;
+            lock (this.lockObject)
+            {
+                res = AssignDownloadTaskToPeerLocked(peer, blockHash);
+            }
+
+            return res;
+        }
+
+        /// <summary>
+        /// Assigns a download task to a specific peer.
+        /// </summary>
+        /// <param name="peer">Peer to be assigned the new task.</param>
+        /// <param name="blockHash">Hash of the block to download from <paramref name="peer"/>.</param>
+        /// <returns><c>true</c> if the block was assigned to the peer, <c>false</c> in case the block has already been assigned to someone.</returns>
+        /// <remarks>The caller of this method is responsible for holding <see cref="lockObject"/>.</remarks>
+        private bool AssignDownloadTaskToPeerLocked(BlockPullerBehavior peer, uint256 blockHash)
```

Comment:
oh yeah, i know what you mean, this is the result of refactoring that i did have another method calling the locked version and i removed it later (similarly for ReleaseDownloadTaskAssignmentLocked, but in that case there really are two methods to call it, so that one needs to stay)

so yes, this one i fix, it is not needed to be separated, tx